mlir/test/lib/Dialect/Rock/TestTransformationMapsUtils.cpp

```diff
@@ -0,0 +1,203 @@
+        rock::untransform(builder, testValue);
+
+    auto removeDims = RequestProcessor::getTests(testHandleOp);
+    FailureOr<ArrayAttr> maybeNewArrayAttr =
+        rock::removeUpperDims(builder, transformAttrs, removeDims);
+
+    return maybeNewArrayAttr;
+  }
+
+  LogicalResult matchAndRewrite(func::FuncOp func,
+                                PatternRewriter &b) const final {
+
+    FailureOr<Operation *> maybeTestHandleOp = getTestHandleOp(func);
+    if (failed(maybeTestHandleOp)) {
+      return failure();
+    }
+    Operation *testHandleOp = maybeTestHandleOp.value();
+
+    // remove indices
+    FailureOr<ArrayAttr> maybeNewTrMapAttrs;
+    if (testHandleOp->getAttr("names_to_drop")) {
+      maybeNewTrMapAttrs = removeUpperDims<ByNames>(b, testHandleOp);
+    } else if (testHandleOp->getAttr("indices_to_drop")) {
+      maybeNewTrMapAttrs = removeUpperDims<ByIndices>(b, testHandleOp);
+    } else {
+      emitError(UnknownLoc::get(func.getContext()),
+                "`remove_dims` op does not have `drop` attr");
+      return failure();
+    }
+
+    if (failed(maybeNewTrMapAttrs)) {
+      return failure();
+    }
+    ArrayAttr newTrMapAttrs = maybeNewTrMapAttrs.value();
+
+    // construct a new function signature and remove the old function body
+    Attribute attr = *(std::prev(newTrMapAttrs.end()));
+    auto firstTrMap = attr.cast<rock::TransformMapAttr>();
+    DenseI64ArrayAttr firstLowerBounds = firstTrMap.getLowerBounds();
+
+    Type inputType = func.getArgument(0).getType();
+    Type inputElementType = inputType.cast<MemRefType>().getElementType();
+
+    auto newInputType =
+        MemRefType::get(firstLowerBounds.asArrayRef(), inputElementType);
+    auto newFuncType =
+        FunctionType::get(func->getContext(), {newInputType}, {});
+
+    func.eraseBody();
+    func.setFunctionType(newFuncType);
+
+    Block *newEntryBlock = func.addEntryBlock();
+    b.setInsertionPointToStart(newEntryBlock);
+
+    // insert new transform ops
+    Location loc = b.getUnknownLoc();
+    Value input = newEntryBlock->getArgument(0);
+    for (auto trMapAttr : llvm::reverse(newTrMapAttrs)) {
+      auto trOp = b.create<rock::TransformOp>(
+          loc, input, trMapAttr.cast<TransformMapAttr>());
+      input = trOp.getOutput();
+    }
+    b.create<func::ReturnOp>(loc);
+
+    return success();
+  }
+};
+
+class TransformMapRewriter : public PatternRewriter {
+public:
+  TransformMapRewriter(MLIRContext *ctx) : PatternRewriter(ctx) {}
+};
+
+void TransformMapsUtilsTestPass::runOnOperation() {
+  MLIRContext *ctx = getOperation()->getContext();
+
+  RewritePatternSet patterns(ctx);
+  patterns.add<RemoveDimTestPattern>(ctx);
+  FrozenRewritePatternSet frozenPatternSet(std::move(patterns));
+
+  PatternApplicator applicator(frozenPatternSet);
+  applicator.applyDefaultCostModel();
+
+  TransformMapRewriter rewriter(ctx);
+  LogicalResult result = applicator.matchAndRewrite(getOperation(), rewriter);
```

Comment:
Huh, `PatternApplicator` is new. I always just went with the FoldGreedily thingy?. Seems useful